cloud/validations.go

```diff
@@ -0,0 +1,279 @@
+		"storage-endpoint":  map[string]interface{}{"type": "string"},
+		"config":            map[string]interface{}{"type": "object"},
+		"regions":           regionsSchema,
+		"region-config":     map[string]interface{}{"type": "object"},
+	},
+	"additionalProperties": false,
+}
+
+var regionsSchema = map[string]interface{}{
+	"type": "object",
+	"additionalProperties": map[string]interface{}{
+		"type": "object",
+		"properties": map[string]interface{}{
+			"endpoint":          map[string]interface{}{"type": "string"},
+			"identity-endpoint": map[string]interface{}{"type": "string"},
+			"storage-endpoint":  map[string]interface{}{"type": "string"},
+		},
+		"additionalProperties": false,
+	},
+}
+
+// ValidateCloudSet reports any erroneous properties found in cloud metadata
+// YAML. If there are no erroneous properties, then ValidateCloudSet returns nil
+// otherwise it return an error listing all erroneous properties and possible
+// suggestion.
+func ValidateCloudSet(data []byte) error {
+	return validateCloud(data, &cloudSetSchema)
+}
+
+// ValidateOneCloud is like ValidateCloudSet but validates the metadata for only
+// one cloud and not multiple.
+func ValidateOneCloud(data []byte) error {
+	return validateCloud(data, &cloudSchema)
+}
+
+func validateCloud(data []byte, jsonSchema *map[string]interface{}) error {
+	var body interface{}
+	if err := yaml.Unmarshal(data, &body); err != nil {
+		return errors.Annotate(err, "cannot unmarshal yaml cloud metadata")
+	}
+
+	jsonBody := yamlToJSON(body)
+	invalidKeys, err := validateCloudMetaData(jsonBody, jsonSchema)
+
+	if err != nil {
+		return errors.Annotate(err, "cannot validate yaml cloud metadata")
+	}
+
+	formatKeyError := func(invalidKey, similarValidKey string) string {
+		str := fmt.Sprintf("property %s is invalid.", invalidKey)
+		if similarValidKey != "" {
+			str = fmt.Sprintf("%s Perhaps you mean \"%s\".", str, similarValidKey)
+		}
+		return str
+	}
+
+	cloudValidationError := ValidationWarning{}
+	for k, v := range invalidKeys {
+		cloudValidationError.Messages = append(cloudValidationError.Messages, formatKeyError(k, v))
+	}
+
+	if len(cloudValidationError.Messages) != 0 {
+		return &cloudValidationError
+	}
+
+	return nil
+}
+
+func cloudTags() []string {
+	keys := make(map[string]struct{})
+	collectTags(reflect.TypeOf((*cloud)(nil)), "yaml", []string{"map[string]*cloud.region", "yaml.MapSlice"}, &keys)
+	keyList := make([]string, 0, len(keys))
+	for k := range keys {
+		keyList = append(keyList, k)
+	}
+
+	return keyList
+}
+
+// collectTags returns a set of keys for a specified struct tag. If no tag is
+// specified for a particular field of the argument struct type, then the
+// all-lowercase field name is used as per Go tag conventions. If the tag
+// specified is not the name a conventionally formatted go struct tag, then the
+// results of this function are invalid. Values of invalid kinds result in no
+// processing.
+
```

Comment:
delete this line, so the doc comment is attached to the function
(then "go doc -u github.com/juju/juju/cloud.collectTags" will show the comment)